lib/checksum.c

```diff
@@ -42,794 +41,858 @@
 #include "checksum.h"
 
 #include "checksums/blake2/blake2.h"
-#include "checksums/cfarmhash.h"
-#include "checksums/city.h"
-#include "checksums/citycrc.h"
+#include "checksums/highwayhash.h"
+#include "checksums/metrohash.h"
 #include "checksums/murmur3.h"
 #include "checksums/sha3/sha3.h"
-#include "checksums/spooky-c.h"
 #include "checksums/xxhash/xxhash.h"
 
 #include "utilities.h"
 
 #define _RM_CHECKSUM_DEBUG 0
 
+//////////////////////////////////
+//    BUFFER IMPLEMENTATION     //
+//////////////////////////////////
+
+RmBuffer *rm_buffer_new(gsize buf_size) {
+    RmBuffer *self = g_slice_new0(RmBuffer);
+    self->data = g_slice_alloc(buf_size);
+    self->buf_size = buf_size;
+    return self;
+}
+
+void rm_buffer_free(RmBuffer *buf) {
+    g_slice_free1(buf->buf_size, buf->data);
+    g_slice_free(RmBuffer, buf);
+}
+
+static gboolean rm_buffer_equal(RmBuffer *a, RmBuffer *b) {
+    return (a->len == b->len && memcmp(a->data, b->data, a->len) == 0);
+}
+
 ///////////////////////////////////////
-//    BUFFER POOL IMPLEMENTATION     //
+//  RMDIGEST INTERFACE DEFINITIONS   //
 ///////////////////////////////////////
 
-RmOff rm_buffer_size(RmBufferPool *pool) {
-    return pool->buffer_size;
+/* Each digest type must have an RmDigestInterface defined as follows: */
+typedef gpointer (*RmDigestNewFunc)(void);
+typedef void (*RmDigestFreeFunc)(gpointer state);
+typedef void (*RmDigestUpdateFunc)(gpointer state, const unsigned char *data, gsize size);
+typedef gpointer (*RmDigestCopyFunc)(gpointer state);
+typedef void (*RmDigestStealFunc)(gpointer state, guint8 *result);
+typedef guint (*RmDigestLenFunc)(gpointer state);
+
+typedef struct RmDigestInterface {
+    const char *name;           // hash name
+    const guint bits;           // length of the output checksum in bits (if const)
+    RmDigestLenFunc len;        // return length of the output checksum in bytes
+    RmDigestNewFunc new;        // returns new digest->state
+    RmDigestFreeFunc free;      // frees state allocated by new()
+    RmDigestUpdateFunc update;  // hashes data into state
+    RmDigestCopyFunc copy;      // allocates and returns a copy of passed state
+    RmDigestStealFunc steal;    // writes checksum (as binary) to *result
+} RmDigestInterface;
```

Comment:
Good work, I like that.

Reply:
I must have had a good teacher (thinking formats.h)